### PR TITLE
[TR#185] Side Menu Update

### PR DESCRIPTION
**Pull request by nexticodes**
[//]: # (Title Template: "[TR_185] Side Menu Update")
---

#### Please check if the PR fulfills these requirements

- [x] The changes don't come from your master branch.

---

#### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
This feature changes the format and layout of the side menu that appears when the hamburger button is clicked by the user. The user's name has been moved from being below the avatar on the opposite side (above). New menu options have been added while some of the current menu options have been disabled along with the lines separating main menu options as per the Figma mockup.

#### What is the current behavior? (You can also link to an open issue here)
Client-Side Menu currently contains 5 options -- My Claims, My Profile, Settings, Help and Log Out.
Donor-Side Menu has 6 options -- Scan QR Code, My Donations, My Profile, Settings, Help and Log out.
Both menus still have the user's name below the avatar and some of its options lead the user back to the Dashboard.

#### What is the new behavior? (if this is a feature change)
No new behavior, other than the changes in the number of options and the placement of the username.
New Client-Side Menu now contains 6 options -- Donations, Claims, Pending, History, Contact Us, and Logout.
New Donor-Side Menu also has 6 options -- Scan QR Code, Donations, Active, History, Contact Us and Logout.

#### Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)

No breaking changes.

#### Other information

None!

Ready to make a change whenever! Thank you all for reviewing :)

---

#### Images (before/ after screenshots, interaction GIFs, ...)
![Before/after example image]

##### Client

SideMenu_Client_BEFORE

##### Donor
SideMenu_Donor_BEFORE

##### Client
SideMenu_Client_AFTER

##### Donor
SideMenu_Donor_AFTER

### CODE REVIEW

File: src/routes/Route.tsx [hunk 1]
```diff
@@ -54,56 +47,72 @@ const donorOrClientDrawer = () => {
 		QRCodeScannerScreen: {
 			screen: QRCodeScannerScreen,
 			navigationOptions: {
-				drawerLabel: 'Scan QR Code',
-				drawerIcon: DrawerIcon('qrCode'),
+				drawerLabel: <MainMenu text="Scan QR Code" icon="qrCode" />,
```
Comment (renburnett) [inline src/routes/Route.tsx:50]: nifty refactor!

File: src/elements/MenuDrawer/MainMenu/MainMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import * as colors from '@util/colors';
+
+import styles from './MainMenu.styles';
+
+const MainMenu = props => {
+	let menuStyle = {};
+	if (props.text === 'Contact Us') {
+		menuStyle = {
+			...styles.mainMenu,
+			marginBottom: 50,
```
Comment (renburnett) [inline src/elements/MenuDrawer/MainMenu/MainMenu.tsx]: Not really an issue, but perhaps we could stick these styles in an object in the styles file and spread them alongside the `mainMenu` styles? 
e.g.
```
menuStyle = {
    ...styles.mainMenu,
    ...styles.contactUs,
}
```

Just a thought 👍

File: src/elements/MenuDrawer/SubMenu/SubMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import React from 'react';
+import { Text } from 'react-native';
+
+import styles from './SubMenu.styles';
+
+const SubMenu = props => <Text style={{ ...styles.subMenu, marginBottom: 10 }}>{props.text}</Text>;
```
Comment (renburnett) [inline src/elements/MenuDrawer/SubMenu/SubMenu.tsx]: Did you mean to leave the `marginBottom` styling outside of the `subMenu` obj?

File: src/elements/ScrollContainer/EndOfScrollWrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+export default function EndOfScrollWrapper({
+	onScrollToEnd,
+	style,
+	children,
+}: EndOfScrollWrapperProps) {
+	const isCloseToBottom = ({ layoutMeasurement, contentOffset, contentSize }) => {
+		const paddingToBottom = 20;
+		return layoutMeasurement.height + contentOffset.y >= contentSize.height - paddingToBottom;
```
Comment (renburnett) [inline src/elements/ScrollContainer/EndOfScrollWrapper.tsx]: nifty!

File: src/routes/Route.tsx [hunk 5]
```diff
@@ -54,56 +47,72 @@ const donorOrClientDrawer = () => {
...
+				drawerLabel: <MainMenu text="Contact Us" icon="help" />,
 			},
 		},
+
 		LogoutScreen: {
 			screen: LogoutScreen,
 			navigationOptions: {
-				drawerLabel: 'Log Out',
-				drawerIcon: DrawerIcon('logout'),
+				drawerLabel: <MainMenu text="Log out" icon="logout" />,
```
Comment (renburnett) [inline src/routes/Route.tsx]: very very very small thing but the "out" in `Log out` was capitalized before this PR 
:)

**Review by renburnett — DISMISSED** [review]
Tested on iphone emulator and it works 👍 
I have a few comments/questions below but other than that, everything looks great!

File: src/elements/ScrollContainer/EndOfScrollWrapper.tsx (hunk 4 above)
Comment (nexticodes) [reply src/elements/ScrollContainer/EndOfScrollWrapper.tsx]: Soooo... This one is kind of weird. I'm not sure why my commit history includes @CMChemist 's commits.

File: src/elements/MenuDrawer/SubMenu/SubMenu.tsx (hunk 3 above)
Comment (nexticodes) [reply src/elements/MenuDrawer/SubMenu/SubMenu.tsx]: Oh no, I did not! Oops, I'll include it into subMenu 😄

File: src/elements/MenuDrawer/MainMenu/MainMenu.tsx (hunk 2 above)
Comment (nexticodes) [reply src/elements/MenuDrawer/MainMenu/MainMenu.tsx]: Good idea -- to keep everything in one place.

**Comment by duskcloudxu** [comment]
Tested and it works well, are you ready to merge it?

**Review by duskcloudxu — DISMISSED** [review]

**Comment by nexticodes** [comment]
Had to do some cleaning up on the code before merging. Should be good to go still!

**Comment by duskcloudxu** [comment]
There are some problems on IOS simulator...
image

Could you reproduce that on android or web browser?

**Comment by nexticodes** [comment]
I apologize for the late response. I commented yesterday but it didn't go through for some reason.

Anyway, this is due to the styling I have because I have it styled specific to my Android device that has a smaller screen compared to the iPhones.

Will look for alternatives and will push.

**Comment by nexticodes** [comment]
@duskcloudxu @renburnett I have made a change but uncertain in which route I should go for.

I have two options for this so far:
1) (Currently applied) create a filler option that does not render any texts but is only blank and its purpose is to create space between Contact Us and the Log Out buttons.
2) Add marginBottom for Contact Us button and marginTop for Log Out button.

Which one..? 🤔

**Comment by nexticodes** [comment]
Oops, I did not realize that more changes have been made to the Figma. Here's what the updated version should look like!

SideMenu_Donor 6112020 update

File: src/elements/MenuDrawer/MainMenu/MainMenu.styles.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import { StyleSheet, Dimensions } from 'react-native';
+import * as colors from '@util/colors';
+
+// Get Window Height;
+// Used for placement of LogOut button.
+const wh = Dimensions.get('window').height;
+
+console.log(wh);
```
Comment (renburnett) [inline src/elements/MenuDrawer/MainMenu/MainMenu.styles.tsx]: oops let's remove `console.log`

File: src/elements/MenuDrawer/MenuDrawer.styles.tsx [hunk 7]
```diff
@@ -13,17 +15,20 @@ export default StyleSheet.create({
 	container: {
 		flex: 1,
 		backgroundColor: colors.NAVY_BLUE,
-		alignItems: 'flex-end',
+		marginTop: -20,
+		flexDirection: 'column',
+		justifyContent: 'space-between',
+		height: vh - (vh / 3),
```
Comment (renburnett) [inline src/elements/MenuDrawer/MenuDrawer.styles.tsx:21]: nifty media query and math!

File: src/elements/MenuDrawer/MainMenu/MainMenu.styles.tsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import { StyleSheet, Dimensions } from 'react-native';
+import * as colors from '@util/colors';
+
+// Get Window Height;
+// Used for placement of LogOut button.
+const wh = Dimensions.get('window').height;
```
Comment (renburnett) [inline src/elements/MenuDrawer/MainMenu/MainMenu.styles.tsx]: is this being used?

File: src/elements/MenuDrawer/MainMenu/MainMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+import { Icon } from '@elements';
+import { IconName } from '@elements/Icon';
+import * as colors from '@util/colors';
+
+import styles from './MainMenu.styles';
+
+const MainMenu = props => {
+	let menuStyle = {};
+	const { icon, text } = props;
```
Comment (renburnett) [inline src/elements/MenuDrawer/MainMenu/MainMenu.tsx:12]: not a big deal but you can save a line but destructuring the `icon` and `text` in place of `props` 
i.e. `const MainMenu = ({ icon, text }) => {`
also I think our linter prefers it

File: src/routes/Route.tsx [hunk 10]
```diff
@@ -1,14 +1,12 @@
 import React from 'react';
+import { View } from 'react-native';
```
Comment (renburnett) [inline src/routes/Route.tsx:2]: is this being used?

**Comment by renburnett** [comment]
> @duskcloudxu @renburnett I have made a change but uncertain in which route I should go for.
> 
> I have two options for this so far:
> 
>     1. (Currently applied) create a filler option that does not render any texts but is only blank and its purpose is to create space between Contact Us and the Log Out buttons.
> 
>     2. Add marginBottom for Contact Us button and marginTop for Log Out button.
> 
> 
> Which one..? 🤔

I apologize for the late reply-- I think your solution of adding a filler/spacer element makes the most sense. 👍  
I think it will make it easier to add other menu items in the future-- instead of having to modify both `ContactUs` and `LogOut` margins when we add new items, we'll just remove or modify the `SpacerElement`.

---

- So I tested this on my iOS emulator and my physical iPhone 6 and both work well for me!
- I had a few small comments about some lines that should probably be removed
- Also! Would you mind `rebasing` to squash all of these 25 commits into one?

File: src/elements/MenuDrawer/MainMenu/MainMenu.tsx (hunk 9 above)
Comment (nexticodes) [reply src/elements/MenuDrawer/MainMenu/MainMenu.tsx:12]: Oh I didn't think of doing that! Good catch 👍

**Comment by nexticodes** [comment]
Closing because I created a new one [HERE](https://github.com/FoodIsLifeBGP/banana-rn/pull/103).

**Pull request closed without merging**